pkg/handler/collector/oci/oci.go

```diff
@@ -182,109 +210,286 @@ func (o *ociCollector) getTagsAndFetch(ctx context.Context, repo string, tags []
+			if _, ok := wellKnownOCIArtifactTypes[referrerDesc.ArtifactType]; ok {
+				referrerDescDigest := referrerDesc.Digest.String()
+
+				if !o.isDigestCollected(repo, referrerDescDigest) {
+					logger.Infof("Fetching referrer %s with artifact type %s", referrerDescDigest, referrerDesc.ArtifactType)
+					referrerDigest := fmt.Sprintf("%v@%v", repo, referrerDescDigest)
+					e := fetchOCIArtifactBlobs(ctx, rc, referrerDigest, referrerDesc.ArtifactType, docChannel)
+					if e != nil {
+						errorChan <- fmt.Errorf("failed retrieving artifact blobs from registry: %w", err)
+					}
+					o.markDigestAsCollected(repo, referrerDescDigest)
 				}
-				docChannel <- doc
+			} else {
+				logger.Infof("Skipping referrer %s with unknown artifact type %s", referrerDesc.Digest, referrerDesc.ArtifactType)
 			}
-			o.checkedDigest[repo] = append(o.checkedDigest[repo], digestTag)
+
+		}(referrerDesc)
+	}
+
+	// Wait for all goroutines to finish
+	wg.Wait()
+
+	// Close the errorChan to signal that all errors have been collected
+	close(errorChan)
+
+	// Check if any errors occurred during processing
+	for err := range errorChan {
+		if err != nil {
+			return err // Return the first error encountered
 		}
 	}
 
 	return nil
 }
 
-func contains(elems []string, v string) bool {
-	for _, s := range elems {
-		if v == s {
-			return true
+// fetchOCIArtifactBlobs fetches the blobs of an OCI artifact and sends them to the provided docChannel.
+// It takes a context.Context, a *regclient.RegClient, an artifact string, an artifactType string, and a docChannel chan<- *processor.Document as input.
+// Note that we are not concurrently fetching the layers since we will usually have 1 layer per artifact.
+// It returns an error if there was an issue fetching the artifact blobs.
+func fetchOCIArtifactBlobs(ctx context.Context, rc *regclient.RegClient, artifact string, artifactType string, docChannel chan<- *processor.Document) error {
+	logger := logging.FromContext(ctx)
+	r, err := ref.New(artifact)
+	if err != nil {
+		return fmt.Errorf("unable to parse OCI reference: %v", artifact)
+	}
+
+	m, err := rc.ManifestGet(ctx, r)
+	if err != nil {
+		// this is a normal behavior, not an error when the digest does not have an attestation
+		// explicitly logging it as info to avoid call-stack when logging
+		logger.Infof("unable to get manifest for %v: %v", artifact, err)
+		return nil
+	}
+
+	// go through layers in reverse
+	mi, ok := m.(manifest.Imager)
+	if !ok {
+		return fmt.Errorf("reference is not a known image media type")
+	}
+	layers, err := mi.GetLayers()
+	if err != nil {
+		return err
+	}
+	for i := len(layers) - 1; i >= 0; i-- {
+		if ctx.Err() != nil {
+			return ctx.Err()
 		}
+		blob, err := rc.BlobGet(ctx, r, layers[i])
+		if err != nil {
+			return fmt.Errorf("failed pulling layer %d: %w", i, err)
+		}
+		defer blob.Close()
+		btr1, err := blob.RawBody()
+		if err != nil {
+			return fmt.Errorf("failed reading layer %d: %w", i, err)
+		}
```

Comment:
Defer inside a for loop can be problematic if there are lots of layers. 
```suggestion
		btr1, err := blob.RawBody()
		closeErr := blob.Close()
		if err != nil {
			return fmt.Errorf("failed reading layer %d: %w", i, err)
		}
		if closeErr != nil {
			return fmt.Errorf("failed close layer %d: %w", i, err)
		}
```